File: crates/client/src/facades/instance.rs
use std::time::Duration;

use ipc::errors::IpcError;
use proto::content::{
    ContentAddSpec, ContentFailure, ContentKind, ContentUpdate, InstalledContent,
    InstanceContentAdd, InstanceContentAddParams, InstanceContentCheckUpdates,
    InstanceContentCheckUpdatesParams, InstanceContentEnable, InstanceContentEnableParams,
    InstanceContentList, InstanceContentListParams, InstanceContentRemove,
    InstanceContentRemoveParams, InstanceContentSetVersion, InstanceContentSetVersionParams,
    InstanceContentUpdate, InstanceContentUpdateParams, UntrackedFile,
};
use proto::instance::{AddressPing, AddressPingParams};
use proto::instance::{
    InstanceConfigGet, InstanceConfigGetParams, InstanceConfigList, InstanceConfigSet,
    InstanceConfigSetParams, InstanceCreate, InstanceCreateParams, InstanceDetails,
    InstanceFlavors, InstanceInfo, InstanceInfoQuery, InstanceLaunch, InstanceLaunchDoneEvent,
    InstanceLaunchParams, InstanceList, InstanceLoaders, InstanceLogs, InstanceLogsParams,
    InstanceProfileCapture, InstanceProfileCreate, InstanceProfileCreateParams,
    InstanceProfileEdit, InstanceProfileEditParams, InstanceProfileList, InstanceProfileRef,
    InstanceProfileRelease, InstanceProfileRemove, InstanceProfileRename,
    InstanceProfileRenameParams, InstanceProfileUse, InstanceRef, InstanceRemove, InstanceRename,
    InstanceRenameParams, InstanceResolve, InstanceServerEdit, InstanceServerEditParams,
    InstanceServerMove, InstanceServerMoveParams, InstanceServerRef, InstanceServerRemove,
    InstanceServers, InstanceServersWriteResult, InstanceStop, InstanceStopParams, InstanceUpdate,
    InstanceUpdateParams, InstanceVersions, InstanceWorlds, Profile, QuickPlay, ServerEntry,
    WorldInfo,
};
use proto::minecraft::{
    ConfigEntry, Flavor, GameVersion, InstanceProfile, LoadersParams, ProvisionProgress,
    ResolveParams, VersionsParams,
};
use proto::process::ProcessLogLine;
use proto::server::ServerPingResult;

use crate::facades::jobs::{forward, run_content_job};
use crate::session::{job_id, Session};

pub struct Instance<'a> {
    pub(crate) session: &'a Session,
}

impl Instance<'_> {
    pub async fn flavors(&self) -> Result<Vec<Flavor>, IpcError> {
        Ok(self
            .session
            .call::<InstanceFlavors>(&proto::Empty {})
            .await?
            .flavors)
    }

    pub async fn versions(&self, flavor: &str) -> Result<Vec<GameVersion>, IpcError> {
        let params = VersionsParams {
            flavor: flavor.to_string(),
        };
        Ok(self
            .session
            .call::<InstanceVersions>(&params)
            .await?
            .versions)
    }

    pub async fn resolve(
        &self,
        flavor: &str,
        version: &str,
        loader_version: Option<String>,
    ) -> Result<InstanceProfile, IpcError> {
        let params = ResolveParams {
            flavor: flavor.to_string(),
            version: version.to_string(),
            loader_version,
        };
        self.session.call::<InstanceResolve>(&params).await
    }

    pub async fn loaders(&self, flavor: &str, version: &str) -> Result<Vec<String>, IpcError> {
        let params = LoadersParams {
            flavor: flavor.to_string(),
            version: version.to_string(),
        };
        Ok(self.session.call::<InstanceLoaders>(&params).await?.loaders)
    }

    /// Create an instance record (the profile is resolved upstream, so this can
    /// take a little while; files are materialised at launch).
    pub async fn create(
        &self,
        name: &str,
        flavor: &str,
        version: &str,
        loader_version: Option<String>,
        config: Vec<ConfigEntry>,
    ) -> Result<InstanceInfo, IpcError> {
        let params = InstanceCreateParams {
            name: name.to_string(),
            flavor: flavor.to_string(),
            version: version.to_string(),
            loader_version,
            config,
        };
        Ok(self
            .session
            .call_with_timeout::<InstanceCreate>(&params, Duration::from_secs(60))
            .await?
            .instance)
    }

    /// Move a stopped instance to another version (the profile is resolved
    /// upstream; the new files materialise at the next launch). Nothing is
    /// backed up first — instances have no backups. `allow_downgrade` asserts
    /// the user confirmed a downgrade.
    pub async fn update(
        &self,
        instance: &str,
        version: &str,
        loader_version: Option<String>,
        allow_downgrade: bool,
    ) -> Result<InstanceInfo, IpcError> {
        let params = InstanceUpdateParams {
            instance: instance.to_string(),
            version: version.to_string(),
            loader_version,
            allow_downgrade,
        };
        Ok(self
            .session
            .call_with_timeout::<InstanceUpdate>(&params, Duration::from_secs(10 * 60))
            .await?
            .instance)
    }

    pub async fn list(&self) -> Result<Vec<InstanceInfo>, IpcError> {
        Ok(self
            .session
            .call::<InstanceList>(&proto::Empty {})
            .await?
            .instances)
    }

    /// The instance's static, informational view (locations + disk footprint).
    pub async fn info(&self, instance: &str) -> Result<InstanceDetails, IpcError> {
        self.session
            .call::<InstanceInfoQuery>(&instance_ref(instance))
            .await
    }

    /// The instance's save worlds, each described from its own `level.dat` —
    /// the worlds a datapack can install into.
    pub async fn worlds(&self, instance: &str) -> Result<Vec<WorldInfo>, IpcError> {
        Ok(self
            .session
            .call::<InstanceWorlds>(&instance_ref(instance))
            .await?
            .worlds)
    }

    /// The instance's multiplayer list (`servers.dat`), in the order the game
    /// shows it.
    pub async fn servers(&self, instance: &str) -> Result<Vec<ServerEntry>, IpcError> {
        Ok(self
            .session
            .call::<InstanceServers>(&instance_ref(instance))
            .await?
            .servers)
    }

    /// Add an entry to the multiplayer list, or rewrite the one `target` names
    /// (empty adds). The running game owns the file, so the result carries a
    /// warning when a session is open to say the edit will be overwritten.
    pub async fn server_edit(
        &self,
        instance: &str,
        target: &str,
        name: &str,
        address: &str,
        accept_textures: bool,
    ) -> Result<InstanceServersWriteResult, IpcError> {
        self.session
            .call::<InstanceServerEdit>(&InstanceServerEditParams {
                instance: instance.to_string(),
                server: target.to_string(),
                name: name.to_string(),
                address: address.to_string(),
                accept_textures,
            })
            .await
    }

    pub async fn server_remove(
        &self,
        instance: &str,
        target: &str,
    ) -> Result<InstanceServersWriteResult, IpcError> {
        self.session
            .call::<InstanceServerRemove>(&InstanceServerRef {
                instance: instance.to_string(),
                server: target.to_string(),
            })
            .await
    }

    /// Move an entry to another slot of the multiplayer list; `position`
    /// counts over the visible entries, which is what a caller has listed.
    pub async fn server_move(
        &self,
        instance: &str,
        target: &str,
        position: u32,
    ) -> Result<InstanceServersWriteResult, IpcError> {
        self.session
            .call::<InstanceServerMove>(&InstanceServerMoveParams {
                instance: instance.to_string(),
                server: target.to_string(),
                position,
            })
            .await
    }

    /// Status of a multiplayer address, over the Server List Ping — what the
    /// in-game list shows beside each entry.
    pub async fn ping_address(&self, address: &str) -> Result<ServerPingResult, IpcError> {
        self.session
            .call::<AddressPing>(&AddressPingParams {
                address: address.to_string(),
            })
            .await
    }

    pub async fn remove(&self, instance: &str) -> Result<(), IpcError> {
        self.session
            .call::<InstanceRemove>(&instance_ref(instance))
            .await?;
        Ok(())
    }

    /// Rename a stopped instance; the id (directory slug) is re-derived from
    /// the new name. Returns the updated record, whose `id` and `name` reflect
    /// the rename.
    pub async fn rename(&self, instance: &str, name: &str) -> Result<InstanceInfo, IpcError> {
        let params = InstanceRenameParams {
            instance: instance.to_string(),
            name: name.to_string(),
        };
        self.session.call::<InstanceRename>(&params).await
    }

    /// Stop a session (when `session` is set) or all of the instance's sessions.
    pub async fn stop(&self, instance: &str, session: Option<String>) -> Result<(), IpcError> {
        let params = InstanceStopParams {
            instance: instance.to_string(),
            session,
        };
        self.session.call::<InstanceStop>(&params).await?;
        Ok(())
    }

    pub async fn logs(
        &self,
        instance: &str,
        session: Option<String>,
        tail: Option<usize>,
    ) -> Result<Vec<ProcessLogLine>, IpcError> {
        let params = InstanceLogsParams {
            instance: instance.to_string(),
            session,
            tail,
        };
        Ok(self.session.call::<InstanceLogs>(&params).await?.lines)
    }

    /// Read one JVM setting; `None` when it is not set (a `not_found` from the
    /// daemon).
    pub async fn config_get(&self, instance: &str, key: &str) -> Result<Option<String>, IpcError> {
        let params = InstanceConfigGetParams {
            instance: instance.to_string(),
            key: key.to_string(),
        };
        Ok(self
            .session
            .try_call::<InstanceConfigGet>(&params)
            .await?
            .map(|r| r.value))
    }

    pub async fn config_set(&self, instance: &str, key: &str, value: &str) -> Result<(), IpcError> {
        let params = InstanceConfigSetParams {
            instance: instance.to_string(),
            key: key.to_string(),
            value: value.to_string(),
        };
        self.session.call::<InstanceConfigSet>(&params).await?;
        Ok(())
    }

    pub async fn config_list(&self, instance: &str) -> Result<Vec<ConfigEntry>, IpcError> {
        Ok(self
            .session
            .call::<InstanceConfigList>(&instance_ref(instance))
            .await?
            .entries)
    }

    /// Launch an instance, blocking until the game process has spawned (or the
    /// preparation failed) and forwarding each progress event to `on_progress`.
    /// `profile` overrides the active content profile for this launch (empty
    /// keeps it; `none` launches with no profile); `quick_play` joins a world or
    /// server on start instead of opening to the title screen. Returns the
    /// supervised process id and pid, plus anything the preparation could not do
    /// properly (a sync target left unshared, for instance).
    pub async fn launch(
        &self,
        instance: &str,
        account: &str,
        new_session: bool,
        profile: &str,
        quick_play: Option<QuickPlay>,
        on_progress: impl Fn(&ProvisionProgress) + Send + Sync + 'static,
    ) -> Result<InstanceLaunchDoneEvent, IpcError> {
        let id = job_id("instance-launch");
        let session = self.session;
        let params = InstanceLaunchParams {
            instance: instance.to_string(),
            account: account.to_string(),
            id: id.clone(),
            new_session,
            profile: profile.to_string(),
            quick_play,
        };
        let payload = self
            .session
            .run_job(
                &id,
                "instance.launch.done",
                "instance.launch.error",
                forward(on_progress),
                move || async move { session.call::<InstanceLaunch>(&params).await.map(|_| ()) },
            )
            .await?;

        serde_json::from_value(payload).map_err(|e| IpcError::Malformed(e.to_string()))
    }

    /// Install a batch of content into an instance, blocking until the daemon
    /// reports done or error and forwarding each progress event to
    /// `on_progress`. Returns everything installed (items plus required
    /// dependencies) and, per item that could not be installed, a failure.
    pub async fn content_add(
        &self,
        instance: &str,
        spec: ContentAddSpec,
        on_progress: impl Fn(&ProvisionProgress) + Send + Sync + 'static,
    ) -> Result<(Vec<InstalledContent>, Vec<ContentFailure>), IpcError> {
        let id = job_id("instance-content-add");
        let params = InstanceContentAddParams {
            instance: instance.to_string(),
            spec,
            id: id.clone(),
        };
        let session = self.session;
        run_content_job(session, &id, on_progress, move || async move {
            session
                .call::<InstanceContentAdd>(&params)
                .await
                .map(|_| ())
        })
        .await
    }

    /// The instance's active profile name (empty = none) and every profile.
    pub async fn profiles(&self, instance: &str) -> Result<(String, Vec<Profile>), IpcError> {
        let result = self
            .session
            .call::<InstanceProfileList>(&instance_ref(instance))
            .await?;
        Ok((result.active, result.profiles))
    }

    /// Create a profile — seeded from the pool by default, or empty.
    pub async fn create_profile(
        &self,
        instance: &str,
        name: &str,
        seed_from_pool: bool,
    ) -> Result<Profile, IpcError> {
        let params = InstanceProfileCreateParams {
            instance: instance.to_string(),
            name: name.to_string(),
            seed_from_pool,
        };
        self.session.call::<InstanceProfileCreate>(&params).await
    }

    /// Removing the active profile clears the active selection.
    pub async fn remove_profile(&self, instance: &str, name: &str) -> Result<(), IpcError> {
        self.session
            .call::<InstanceProfileRemove>(&profile_ref(instance, name))
            .await?;
        Ok(())
    }

    pub async fn rename_profile(
        &self,
        instance: &str,
        name: &str,
        new_name: &str,
    ) -> Result<Profile, IpcError> {
        let params = InstanceProfileRenameParams {
            instance: instance.to_string(),
            name: name.to_string(),
            new_name: new_name.to_string(),
        };
        self.session.call::<InstanceProfileRename>(&params).await
    }

    /// Set the active profile (empty clears it); applied at the next launch.
    pub async fn use_profile(&self, instance: &str, name: &str) -> Result<(), IpcError> {
        self.session
            .call::<InstanceProfileUse>(&profile_ref(instance, name))
            .await?;
        Ok(())
    }

    /// Add/remove members by pool reference (project id, slug, filename, or
    /// title).
    pub async fn edit_profile(
        &self,
        instance: &str,
        name: &str,
        add: Vec<String>,
        remove: Vec<String>,
    ) -> Result<Profile, IpcError> {
        let params = InstanceProfileEditParams {
            instance: instance.to_string(),
            name: name.to_string(),
            add,
            remove,
        };
        self.session.call::<InstanceProfileEdit>(&params).await
    }

    /// Capture the profile's own settings store (snapshotted from the global
    /// one); launches under it then sync settings against the captured store.
    /// The instance must be stopped.
    pub async fn capture_profile(&self, instance: &str, name: &str) -> Result<(), IpcError> {
        self.session
            .call::<InstanceProfileCapture>(&profile_ref(instance, name))
            .await?;
        Ok(())
    }

    /// Delete the profile's captured store; it inherits the global store
    /// again. The instance must be stopped.
    pub async fn release_profile(&self, instance: &str, name: &str) -> Result<(), IpcError> {
        self.session
            .call::<InstanceProfileRelease>(&profile_ref(instance, name))
            .await?;
        Ok(())
    }

    /// Apply a global profile into the instance's pool — a content job:
    /// references not already present install at their newest compatible
    /// version, tagged with the profile; incompatible ones come back as
    /// failures. Applying never removes de-listed content.
    pub async fn apply_profile(
        &self,
        instance: &str,
        profile: &str,
        on_progress: impl Fn(&ProvisionProgress) + Send + Sync + 'static,
    ) -> Result<(Vec<InstalledContent>, Vec<ContentFailure>), IpcError> {
        let id = job_id("profile-apply");
        let params = proto::profile::InstanceProfileApplyParams {
            instance: instance.to_string(),
            profile: profile.to_string(),
            id: id.clone(),
        };
        let session = self.session;
        run_content_job(session, &id, on_progress, move || async move {
            session
                .call::<proto::profile::InstanceProfileApply>(&params)
                .await
                .map(|_| ())
        })
        .await
    }

    pub async fn content_list(
        &self,
        instance: &str,
        kind: ContentKind,
    ) -> Result<(Vec<InstalledContent>, Vec<UntrackedFile>), IpcError> {
        let params = InstanceContentListParams {
            instance: instance.to_string(),
            kind,
        };
        let result = self.session.call::<InstanceContentList>(&params).await?;
        Ok((result.items, result.untracked))
    }

    /// Uninstall the named items and every mirror of them, in one call. A
    /// non-empty `worlds` instead narrows a datapack to the save worlds it
    /// keeps loading in.
    pub async fn content_remove(
        &self,
        instance: &str,
        kind: ContentKind,
        items: &[String],
        worlds: &[String],
    ) -> Result<(), IpcError> {
        let params = InstanceContentRemoveParams {
            instance: instance.to_string(),
            kind,
            items: items.to_vec(),
            worlds: worlds.to_vec(),
        };
        self.session.call::<InstanceContentRemove>(&params).await?;
        Ok(())
    }

    /// Update platform-sourced content to its newest compatible version — the
    /// named items in one job, or every item of the kind when `items` is empty.
    pub async fn content_update(
        &self,
        instance: &str,
        kind: ContentKind,
        items: &[String],
        on_progress: impl Fn(&ProvisionProgress) + Send + Sync + 'static,
    ) -> Result<Vec<InstalledContent>, IpcError> {
        let id = job_id("instance-content-update");
        let params = InstanceContentUpdateParams {
            instance: instance.to_string(),
            kind,
            items: items.to_vec(),
            id: id.clone(),
        };
        let session = self.session;
        run_content_job(session, &id, on_progress, move || async move {
            session
                .call::<InstanceContentUpdate>(&params)
                .await
                .map(|_| ())
        })
        .await
        .map(|(items, _)| items)
    }

    /// Enable or disable one installed item; a non-empty `worlds` scopes a
    /// datapack toggle to those save worlds.
    pub async fn content_enable(
        &self,
        instance: &str,
        kind: ContentKind,
        item: &str,
        enabled: bool,
        worlds: &[String],
    ) -> Result<(), IpcError> {
        let params = InstanceContentEnableParams {
            instance: instance.to_string(),
            kind,
            item: item.to_string(),
            enabled,
            worlds: worlds.to_vec(),
        };
        self.session.call::<InstanceContentEnable>(&params).await?;
        Ok(())
    }

    /// Which platform-sourced items of the kind have a newer compatible version.
    pub async fn content_check_updates(
        &self,
        instance: &str,
        kind: ContentKind,
    ) -> Result<Vec<ContentUpdate>, IpcError> {
        let params = InstanceContentCheckUpdatesParams {
            instance: instance.to_string(),
            kind,
        };
        let result = self
            .session
            .call_with_timeout::<InstanceContentCheckUpdates>(&params, Duration::from_secs(120))
            .await?;
        Ok(result.updates)
    }

    /// Re-pin one item to a specific published `version` (id or number).
    pub async fn content_set_version(
        &self,
        instance: &str,
        kind: ContentKind,
        item: &str,
        version: &str,
        on_progress: impl Fn(&ProvisionProgress) + Send + Sync + 'static,
    ) -> Result<Vec<InstalledContent>, IpcError> {
        let id = job_id("instance-content-set-version");
        let params = InstanceContentSetVersionParams {
            instance: instance.to_string(),
            kind,
            item: item.to_string(),
            version: version.to_string(),
            id: id.clone(),
        };
        let session = self.session;
        run_content_job(session, &id, on_progress, move || async move {
            session
                .call::<InstanceContentSetVersion>(&params)
                .await
                .map(|_| ())
        })
        .await
        .map(|(items, _)| items)
    }
}

fn instance_ref(instance: &str) -> InstanceRef {
    InstanceRef {
        instance: instance.to_string(),
    }
}

fn profile_ref(instance: &str, name: &str) -> InstanceProfileRef {
    InstanceProfileRef {
        instance: instance.to_string(),
        name: name.to_string(),
    }
}

File: frontend/src/api/instance.ts
/**
 * The `instance.*` channels. Unlike a server, an instance is a lightweight
 * record at create — its files materialise during the `launch` job. An
 * instance can run several concurrent sessions (opt-in via `new_session`);
 * `stop` and `logs` take an optional session id.
 */

import type { ContentAddInput } from './content';
import { call, tryCall } from './core/ipc';
import { type JobRun, runJob } from './core/jobs';
import type {
  ContentDoneEvent,
  ContentKind,
  ContentListResult,
  ContentUpdate,
} from './types/content';
import type {
  InstanceCreateParams,
  InstanceDetails,
  InstanceInfo,
  InstanceLaunchDoneEvent,
  InstanceLaunchParams,
  InstanceProfileListResult,
  InstanceServerEditParams,
  InstanceServersWriteResult,
  InstanceUpdateParams,
  Profile,
  ServerEntry,
  WorldInfo,
} from './types/instance';
import type {
  ConfigEntry,
  Flavor,
  GameVersion,
  InstanceProfile,
  ResolveParams,
} from './types/minecraft';
import type { ProcessLogLine } from './types/process';
import type { ServerPingResult } from './types/server';

export async function flavors(): Promise<Flavor[]> {
  const result = await call<{ flavors: Flavor[] }>('instance.flavors');
  return result.flavors;
}

export async function versions(flavor: string): Promise<GameVersion[]> {
  const result = await call<{ versions: GameVersion[] }>('instance.versions', {
    flavor,
  });
  return result.versions;
}

export function resolve(params: ResolveParams): Promise<InstanceProfile> {
  return call('instance.resolve', params);
}

/** Loader builds for a flavor/version, newest first; empty for vanilla. */
export async function loaders(
  flavor: string,
  version: string,
): Promise<string[]> {
  const result = await call<{ loaders: string[] }>('instance.loaders', {
    flavor,
    version,
  });
  return result.loaders;
}

export async function list(): Promise<InstanceInfo[]> {
  const result = await call<{ instances: InstanceInfo[] }>('instance.list');
  return result?.instances ?? [];
}

/** The instance's static, informational view (locations + disk footprint). */
export function info(instance: string): Promise<InstanceDetails> {
  return call('instance.info', { instance });
}

export async function create(
  params: Partial<InstanceCreateParams>,
): Promise<InstanceInfo> {
  const result = await call<{ instance: InstanceInfo }>(
    'instance.create',
    params,
    { timeoutMs: 60_000 },
  );
  return result.instance;
}

/** The instance pays for the new version at its next launch. */
export async function update(
  params: Omit<InstanceUpdateParams, 'id'>,
): Promise<InstanceInfo> {
  const result = await call<{ instance: InstanceInfo }>(
    'instance.update',
    params,
    { timeoutMs: 10 * 60_000 },
  );
  return result.instance;
}

export function rename(instance: string, name: string): Promise<InstanceInfo> {
  return call('instance.rename', { instance, name });
}

export async function remove(instance: string): Promise<void> {
  await call('instance.remove', { instance });
}

/** The save worlds, each described from its own `level.dat`. */
export async function worlds(instance: string): Promise<WorldInfo[]> {
  const result = await call<{ worlds: WorldInfo[] }>('instance.worlds', {
    instance,
  });
  return result.worlds;
}

/** The multiplayer list (`servers.dat`), in the order the game shows it. */
export async function servers(instance: string): Promise<ServerEntry[]> {
  const result = await call<{ servers: ServerEntry[] }>('instance.servers', {
    instance,
  });
  return result.servers;
}

/**
 * Add an entry to the multiplayer list, or rewrite the one `server` names
 * (empty adds). The running game owns the file, so the result carries a
 * warning when a session is open to say the edit will be overwritten.
 */
export function serverEdit(
  params: Partial<InstanceServerEditParams>,
): Promise<InstanceServersWriteResult> {
  return call('instance.server.edit', params);
}

export function serverRemove(
  instance: string,
  server: string,
): Promise<InstanceServersWriteResult> {
  return call('instance.server.remove', { instance, server });
}

/**
 * Move an entry within the multiplayer list. `position` counts over the
 * visible entries — the rows that were listed — from zero.
 */
export function serverMove(
  instance: string,
  server: string,
  position: number,
): Promise<InstanceServersWriteResult> {
  return call('instance.server.move', { instance, server, position });
}

/** Status of a multiplayer address, over the Server List Ping. */
export function pingAddress(address: string): Promise<ServerPingResult> {
  return call('minecraft.ping', { address });
}

/**
 * Materialise the instance's files and spawn the game as the signed-in
 * account. Resolves once the session is running.
 */
export function launch(
  params: Partial<InstanceLaunchParams>,
  job: JobRun,
): Promise<InstanceLaunchDoneEvent> {
  return runJob<InstanceLaunchDoneEvent>({
    ...job,
    topics: {
      progress: 'instance.launch.progress',
      done: 'instance.launch.done',
      error: 'instance.launch.error',
    },
    start: () => call('instance.launch', { ...params, id: job.id }),
  });
}

/** Stops one named session, or every session of the instance. */
export async function stop(instance: string, session?: string): Promise<void> {
  await call('instance.stop', { instance, session });
}

/** Targets the newest running session unless one is named. */
export async function logs(
  instance: string,
  options: { session?: string; tail?: number } = {},
): Promise<ProcessLogLine[]> {
  const result = await call<{ lines: ProcessLogLine[] }>('instance.logs', {
    instance,
    session: options.session,
    tail: options.tail,
  });
  return result.lines;
}

/** `memory` and `jvm-args` only. */
export const config = {
  async get(instance: string, key: string): Promise<string | null> {
    const result = await tryCall<{ value: string }>('instance.config.get', {
      instance,
      key,
    });
    return result?.value ?? null;
  },

  async set(instance: string, key: string, value: string): Promise<void> {
    await call('instance.config.set', { instance, key, value });
  },

  async list(instance: string): Promise<ConfigEntry[]> {
    const result = await call<{ entries: ConfigEntry[] }>(
      'instance.config.list',
      { instance },
    );
    return result.entries;
  },
};

/**
 * Per-instance content profiles: named selections over the installed pool,
 * enforced by the launch-time mirror reconcile. CRUD applies at the next
 * launch, so it is safe while the instance runs.
 */
export const profiles = {
  list(instance: string): Promise<InstanceProfileListResult> {
    return call('instance.profile.list', { instance });
  },

  /** Seeded with every selectable pool item unless `seedFromPool` is false. */
  create(
    instance: string,
    name: string,
    seedFromPool = true,
  ): Promise<Profile> {
    return call('instance.profile.create', {
      instance,
      name,
      seedFromPool,
    });
  },

  /** Removing the active profile clears the active selection. */
  async remove(instance: string, name: string): Promise<void> {
    await call('instance.profile.remove', { instance, name });
  },

  rename(instance: string, name: string, newName: string): Promise<Profile> {
    return call('instance.profile.rename', {
      instance,
      name,
      newName,
    });
  },

  /** Sets the active profile; an empty `name` clears it. */
  async use(instance: string, name: string): Promise<void> {
    await call('instance.profile.use', { instance, name });
  },

  /**
   * Add/remove members by pool reference (project id, slug, filename, or
   * title); a reference that matches nothing — or only a datapack — errors.
   */
  edit(
    instance: string,
    name: string,
    add: string[] = [],
    remove: string[] = [],
  ): Promise<Profile> {
    return call('instance.profile.edit', { instance, name, add, remove });
  },

  /**
   * Capture the profile's own settings store (snapshotted from the global
   * one); launches under it then sync settings against the captured store.
   * The instance must be stopped.
   */
  async capture(instance: string, name: string): Promise<void> {
    await call('instance.profile.capture', { instance, name });
  },

  /**
   * Delete the profile's captured store; it inherits the global store again.
   * The instance must be stopped.
   */
  async release(instance: string, name: string): Promise<void> {
    await call('instance.profile.release', { instance, name });
  },

  /**
   * Apply a **global** profile into the instance's pool — a content job:
   * references not already present install at their newest compatible
   * version, tagged `profile:<name>`; incompatible ones come back as
   * failures. Applying never removes de-listed content. Refused on a running
   * or busy instance.
   */
  apply(
    instance: string,
    profile: string,
    job: JobRun,
  ): Promise<ContentDoneEvent> {
    return runJob<ContentDoneEvent>({
      ...job,
      topics: { done: 'content.done', error: 'content.error' },
      start: () =>
        call('instance.profile.apply', { instance, profile, id: job.id }),
    });
  },
};

export const content = {
  /** Instances take mods, resourcepacks, shaders, and datapacks. */
  add(
    instance: string,
    spec: ContentAddInput,
    job: JobRun,
  ): Promise<ContentDoneEvent> {
    return runJob<ContentDoneEvent>({
      ...job,
      topics: {
        progress: 'content.progress',
        done: 'content.done',
        error: 'content.error',
      },
      start: () =>
        call('instance.content.add', { instance, ...spec, id: job.id }),
    });
  },

  list(instance: string, kind: ContentKind): Promise<ContentListResult> {
    return call('instance.content.list', { instance, kind });
  },

  /** Uninstalls the named items in one call; a name matching nothing takes none. */
  async remove(
    instance: string,
    kind: ContentKind,
    items: string[],
    worlds: string[] = [],
  ): Promise<void> {
    await call('instance.content.remove', { instance, kind, items, worlds });
  },

  /** `items` empty updates every platform-sourced item of the kind. */
  update(
    instance: string,
    kind: ContentKind,
    items: string[],
    job: JobRun,
  ): Promise<ContentDoneEvent> {
    return runJob<ContentDoneEvent>({
      ...job,
      topics: {
        progress: 'content.progress',
        done: 'content.done',
        error: 'content.error',
      },
      start: () =>
        call('instance.content.update', { instance, kind, items, id: job.id }),
    });
  },

  /** Enable or disable one installed item; applies at the next launch. */
  async enable(
    instance: string,
    kind: ContentKind,
    item: string,
    enabled: boolean,
    worlds: string[] = [],
  ): Promise<void> {
    await call('instance.content.enable', {
      instance,
      kind,
      item,
      enabled,
      worlds,
    });
  },

  /** Which platform items of the kind have a newer compatible version. */
  async checkUpdates(
    instance: string,
    kind: ContentKind,
  ): Promise<ContentUpdate[]> {
    const result = await call<{ updates: ContentUpdate[] }>(
      'instance.content.check_updates',
      { instance, kind },
      { timeoutMs: 120_000 },
    );
    return result.updates;
  },

  /** Re-pin one item to a specific published version (id or number). */
  setVersion(
    instance: string,
    kind: ContentKind,
    item: string,
    version: string,
    job: JobRun,
  ): Promise<ContentDoneEvent> {
    return runJob<ContentDoneEvent>({
      ...job,
      topics: {
        progress: 'content.progress',
        done: 'content.done',
        error: 'content.error',
      },
      start: () =>
        call('instance.content.set_version', {
          instance,
          kind,
          item,
          version,
          id: job.id,
        }),
    });
  },
};
